lib/datadog/tracing/trace_segment.rb

```diff
@@ -17,6 +17,19 @@ class TraceSegment
         :spans,
         :tags
 
+      if RUBY_VERSION < '2.2' # nil.dup only fails in Ruby 2.1
+        # Ensures #initialize can call nil.dup safely
```

Comment:
This comment is incorrect, although the code patterns we're using would probably not trigger the issue.

**That is: on Ruby 2.2 and 2.3 `nil.dup` is still not valid!**

...but the key thing that's "saving" us is that usually we're doing `foo.frozen? ? foo : foo.dup` and because `nil` is frozen on 2.2+, then we would never call `nil.dup` on 2.2/2.3.

Nevertheless, I would suggest -- just in case -- that we:

* Extend this refinement to Ruby < 2.4, so that we don't accidentally try to `nil.dup` on one of those
* Add a test for this behavior (so that this 2.2/2.3 discrepancy would've been caught by CI)

This is because, even though, **right now** we seem in OK shape, maybe next time someone won't remember to check if something is frozen and just dup it.

Reply:
Good point! That's a good catch!